pandas/tests/io/pytables/test_store.py

```diff
@@ -4888,6 +4888,49 @@ def test_unsuppored_hdf_file_error(self, datapath):
         with pytest.raises(ValueError, match=message):
             pd.read_hdf(data_path)
 
+    def test_supported_for_subclasses_dataframe(self):
+        class SubDataFrame(DataFrame):
+            @property
+            def _constructor(self):
+                return SubDataFrame
+
+        data = {"a": [1, 2], "b": [3, 4]}
+        sdf = SubDataFrame(data, dtype=np.intp)
+
+        expected = np.array([[1, 3], [2, 4]], dtype=np.intp)
+
+        with ensure_clean_path("temp.h5") as path:
+            sdf.to_hdf(path, "df")
+            result = read_hdf(path, "df").values
+            tm.assert_numpy_array_equal(result, expected)
+
+        with ensure_clean_path("temp.h5") as path:
+            with HDFStore(path) as store:
+                store.put("df", sdf)
+            result = read_hdf(path, "df").values
+            tm.assert_numpy_array_equal(result, expected)
+
+    def test_supported_for_subclasses_series(self):
+        class SubSeries(Series):
+            @property
+            def _constructor(self):
+                return SubSeries
+
+        sser = SubSeries([1, 2, 3], dtype=np.intp)
+
+        expected = np.array([1, 2, 3], dtype=np.intp)
+
+        with ensure_clean_path("temp.h5") as path:
+            sser.to_hdf(path, "ser")
+            result = read_hdf(path, "ser").values
+            tm.assert_numpy_array_equal(result, expected)
+
+        with ensure_clean_path("temp.h5") as path:
+            with HDFStore(path) as store:
+                store.put("ser", sser)
+            result = read_hdf(path, "ser").values
+            tm.assert_numpy_array_equal(result, expected)
```

Comment:
is it worth leaving TODOs here about switching to `assert_series_equal later`

Reply:
I figured out that the problem with the assertion was that actually ``read_hdf`` would return ``DataFrame``, not ``SubDataFrame``. As a result assertion failed at the stage of comparing types.
Probably need to figure out how one can extract the same subclass of the DataFrame by using ``read_hdf``...
But I guess, this is slightly beyond the scope of the present PR.

So, now I expect that ``read_hdf`` would return ``DataFrame`` with the given values and datatypes and use it for the assertion via ``assert_frame_equal``.